File: src/integrations/telegramFormat.ts
export type Intent = "LONG_OPEN" | "LONG_CLOSE" | "SHORT_OPEN" | "SHORT_CLOSE";

type OrderType = "Market" | "Limit";
type PositionSide = "LONG" | "SHORT";

function emojiAndTitle(intent: Intent | string): { emoji: string; title: string } {
  switch ((intent || "").toUpperCase()) {
    case "LONG_OPEN":
      return { emoji: "🟢", title: "Buy" };
    case "SHORT_OPEN":
      return { emoji: "🔴", title: "Sell" };
    case "LONG_CLOSE":
      return { emoji: "⚪", title: "Close Long" };
    case "SHORT_CLOSE":
      return { emoji: "⚫", title: "Close Short" };
    default:
      return { emoji: "⚪", title: "Signal" };
  }
}

function autoTradeStr(on: boolean): string {
  return on ? "On" : "Off";
}

function tsStr(d?: Date): string {
  const dt = d ?? new Date();
  const pad = (n: number) => n.toString().padStart(2, "0");
  const Y = dt.getFullYear();
  const M = pad(dt.getMonth() + 1);
  const D = pad(dt.getDate());
  const h = pad(dt.getHours());
  const m = pad(dt.getMinutes());
  const s = pad(dt.getSeconds());
  return `${Y}-${M}-${D} ${h}:${m}:${s}`;
}

function formatNumber(value: number | string): string {
  if (typeof value === "number") {
    return value.toFixed(10).replace(/\.0+$/, "").replace(/(\.\d*?)0+$/, "$1").replace(/\.$/, "");
  }
  return value;
}

export function buildSignalMessage(opts: {
  symbol: string;
  intent: Intent;
  orderType?: OrderType;
  positionSide?: PositionSide;
  autoTrade?: boolean;
  leverage?: number;
  marginUSDT?: number;
  quantity?: string;
  reduceOnly?: boolean;
  timestamp?: Date;
}): string {
  const {
    symbol,
    intent,
    orderType = "Market",
    positionSide = "LONG",
    autoTrade = false,
    leverage,
    marginUSDT,
    quantity,
    reduceOnly = false,
    timestamp
  } = opts;

  const { emoji, title } = emojiAndTitle(intent);

  const lines: string[] = [];
  lines.push(`${emoji} SIGNAL - ${title}`);
  lines.push("------------------------");
  lines.push(`Asset: ${symbol}`);

  if (typeof marginUSDT === "number") {
    lines.push(`Margin: ${formatNumber(marginUSDT)} USDT`);
  } else if (quantity) {
    lines.push(`Quantity: ${quantity}`);
  }

  if (typeof leverage === "number") {
    lines.push(`Leverage: ${formatNumber(leverage)}x`);
  }

  lines.push(`Auto-trade: ${autoTradeStr(autoTrade)}`);

  if (intent.includes("CLOSE")) {
    let exitType = orderType;
    if (reduceOnly) exitType += " (Reduce Only)";
    lines.push(`Exit Type: ${exitType}`);
  } else {
    lines.push(`Entry Type: ${orderType}`);
  }

  lines.push(`Position Side: ${positionSide}`);
  lines.push(`Timestamp: ${tsStr(timestamp)}`);

  return lines.join("\n");
}

File: integrations/telegram_format.py
"""Helpers for building Telegram signal messages with a unified layout."""

from __future__ import annotations

from datetime import datetime
from typing import Optional

__all__ = ["build_signal_message"]


def _emoji_and_title(intent: str) -> tuple[str, str]:
    token = (intent or "").upper()
    if token == "LONG_OPEN":
        return "🟢", "Buy"
    if token == "SHORT_OPEN":
        return "🔴", "Sell"
    if token == "LONG_CLOSE":
        return "⚪", "Close Long"
    if token == "SHORT_CLOSE":
        return "⚫", "Close Short"
    return "⚪", "Signal"


def _auto_trade_str(enabled: bool) -> str:
    return "On" if enabled else "Off"


def _timestamp_str(ts: Optional[datetime] = None) -> str:
    ts = ts or datetime.now()
    return ts.strftime("%Y-%m-%d %H:%M:%S")


def _format_number(value: float | int) -> str:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return str(value)
    text = f"{number:.10f}".rstrip("0").rstrip(".")
    return text or "0"


def build_signal_message(
    *,
    symbol: str,
    intent: str,
    order_type: str = "Market",
    position_side: str = "LONG",
    auto_trade: bool = False,
    leverage: Optional[float] = None,
    margin_usdt: Optional[float] = None,
    quantity: Optional[str] = None,
    reduce_only: bool = False,
    timestamp: Optional[datetime] = None,
) -> str:
    emoji, title = _emoji_and_title(intent)

    lines = [f"{emoji} SIGNAL - {title}", "-" * 24]

    lines.append(f"Asset: {symbol}")

    if margin_usdt is not None:
        lines.append(f"Margin: {_format_number(margin_usdt)} USDT")
    elif quantity is not None:
        lines.append(f"Quantity: {quantity}")

    if leverage is not None:
        leverage_text = _format_number(leverage)
        lines.append(f"Leverage: {leverage_text}x")

    lines.append(f"Auto-trade: {_auto_trade_str(auto_trade)}")

    normalised_type = order_type or "Market"
    if "CLOSE" in (intent or "").upper():
        exit_text = normalised_type
        if reduce_only:
            exit_text = f"{exit_text} (Reduce Only)"
        lines.append(f"Exit Type: {exit_text}")
    else:
        lines.append(f"Entry Type: {normalised_type}")

    lines.append(f"Position Side: {position_side.upper()}")

    lines.append(f"Timestamp: {_timestamp_str(timestamp)}")

    return "\n".join(lines)
